python/ray/data/dataset_iterator.py

```diff
@@ -0,0 +1,248 @@
+import abc
+from typing import TYPE_CHECKING, Dict, List, Optional, Union, Iterator
+from typing_extensions import Literal
+
+from ray.data.block import DataBatch
+from ray.util.annotations import PublicAPI
+
+if TYPE_CHECKING:
+    import tensorflow as tf
+    import torch
+    from ray.data._internal.torch_iterable_dataset import TorchTensorBatchType
+
+
+@PublicAPI(stability="beta")
+class DatasetIterator(abc.ABC):
+    """An iterator for reading items from a :class:`~Dataset` or
```

Comment:
Is there any way we can get this excellent docstring to show up in the API docs? I think that the only way is adding a
```
    .. automethod:: __init__
```
to the `autoclass` declaration in the Sphinx manifest, although maybe there's a way for it to pop up on the abstraction landing page in the API docs.

https://ray--31470.org.readthedocs.build/en/31470/data/api/dataset_iterator.html

Reply:
I guess we can just do as https://github.com/ray-project/ray/pull/31470#discussion_r1066397537 w/ `Constructor`.

Reply:
Ah good call, I didn't realize it wasn't getting rendered with the new API TOC. `autoclass` works. Actually we should do the same for `Dataset` and `DatasetPipeline` too, and hide their constructors (I don't think these should be public, right?). I can do that in a follow-up.